tools/checkstyle/checkstyle.bzl

```diff
@@ -0,0 +1,199 @@
+    classpath = ":".join([file.path for file in ctx.files._classpath])
+
+    args = ""
+    inputs = []
+    if ctx.file.config:
+      args += " -c %s" % ctx.file.config.path
+      inputs.append(ctx.file.config)
+    if properties:
+      args += " -p %s" % properties.path
+      inputs.append(properties)
+    if suppressions:
+      inputs.append(suppressions)
+
+    # Build command to run Checkstyle test
+    cmd = " ".join(
+        ["java -cp %s com.puppycrawl.tools.checkstyle.Main" % classpath] +
+        [args] +
+        ["--%s" % x for x in opts] +
+        ["--%s %s" % (k, sopts[k]) for k in sopts] +
+        [file.path for file in ctx.attr.target[JavaSourceFiles].files]
+    )
+
+    # Wrap checkstyle command in a shell script so allow_failure is supported
+    ctx.actions.expand_template(
+        template = ctx.file._checkstyle_sh_template,
+        output = ctx.outputs.checkstyle_script,
+        substitutions = {
+            "{command}" : cmd,
+            "{allow_failure}": str(int(ctx.attr.allow_failure)),
+        },
+        is_executable = True,
+    )
+
+    files = [ctx.outputs.checkstyle_script, ctx.file.license] + ctx.attr.target[JavaSourceFiles].files + ctx.files._classpath + inputs
+    runfiles = ctx.runfiles(
+        files = files,
+        collect_data = True
+    )
+    return DefaultInfo(
+        executable = ctx.outputs.checkstyle_script,
+        files = depset(files),
+        runfiles = runfiles,
+    )
+
+checkstyle_test = rule(
+    implementation = _checkstyle_test_impl,
+    test = True,
+    attrs = {
+        "config": attr.label(
+            allow_single_file=True,
+            doc = "A checkstyle configuration file",
+            default = "//tools/checkstyle:config.xml",
+        ),
+        "suppressions": attr.label(
+            allow_single_file=True,
+            doc = "A checkstyle suppressions file"
+        ),
+        "license": attr.label(
+            allow_single_file=True,
+            doc = "A license file that can be used with the checkstyle license target",
+            default = "//tools/checkstyle:license-header.txt",
+        ),
+        "properties": attr.label(
+            allow_single_file=True,
+            doc = "A properties file to be used"
+        ),
+        "opts": attr.string_list(
+            doc = "Options to be passed on the command line that have no argument"
+        ),
+        "string_opts": attr.string_dict(
+            doc = "Options to be passed on the command line that have an argument"
+        ),
+        "target": attr.label(
+            doc = "The java_library target to check sources on",
+            aspects = [collect_sources],
+            mandatory = True
+        ),
+        "allow_failure": attr.bool(
+            default = False,
+            doc = "Successfully finish the test even if checkstyle failed"
+        ),
+        "_checkstyle_sh_template": attr.label(
+             allow_files = True,
+             single_file = True,
```

Comment:
Docs say about `single_file`:
`Deprecated: Use allow_single_file instead`
https://docs.bazel.build/versions/master/skylark/lib/attr.html#label

BTW, why do we need both `allow_files` and `single_file` and not just `allow_single_file`?

Reply:
@oferb thanks, this is a very valid point!

Reply:
@oferb fixed in b1f8cbdeb78d5810a5d5827a1d198b3e65a7d851